gqcp/src/ONVBasis/SeniorityZeroONVBasis.cpp

```diff
@@ -67,8 +67,35 @@ size_t SeniorityZeroONVBasis::calculateDimension(const size_t K, const size_t N_
  */
 VectorX<double> SeniorityZeroONVBasis::evaluateOperatorDiagonal(const ScalarSQOneElectronOperator<double>& one_op) const {
 
-    const SpinResolvedSelectedONVBasis selected_onv_basis {*this};
-    return selected_onv_basis.evaluateOperatorDiagonal(one_op);
+    // Check if the argument is compatible.
+    const auto K = one_op.dimension();  // number of spatial orbitals
+
+    if (K != this->numberOfSpatialOrbitals()) {
+        throw std::invalid_argument("SeniorityZeroONVBasis::evaluateOperatorDiagonal(const ScalarSQOneElectronOperator<double>&): The number of spatial orbitals for the ONV basis and one-electron operator are incompatible.");
+    }
+
+    // Prepare some variables to be used in the algorithm.
+    const auto N_P = this->numberOfElectronPairs();
+    const auto dim = this->dimension();
+
+    const auto& f = one_op.parameters();
+
+    VectorX<double> diagonal = VectorX<double>::Zero(dim);
+
+
+    // Iterate over every proxy doubly-occupied ONV. Since we are actually using spin-unresolved ONVs, we should multiply contributions by 2.
+    this->forEach([&diagonal, &f](const SpinUnresolvedONV& onv, const size_t I) {
+        double value = 0;  // to be added to the diagonal
+
+        // Loop over occupied orbital index and add the contribution due to the spinor that it occupies.
+        onv.forEach([&value, &f](const size_t p) {
+            value += 2 * f(p, p);  // *2 because of seniority-zero
+        });
+
+        diagonal(I) += value;
+    });
```

Comment:
@guacke @xdvriend @dariatols @nbilliet What do you think of this source code, related to the following older implementation?

https://github.com/lelemmen/GQCP/blob/08cddab57aba3fe3746cf587a9d9b86dee5d9d50/gqcp/src/QCMethod/CI/HamiltonianBuilder/DOCI.cpp#L200-L227

Reply:
There are less for-loops to keep track of in the newer code, so that makes it easier to read. The comment `Iterate over every proxy doubly-occupied ONV.` was not enough for me to explain what `forEach` exactly does, but the body of `forEach` explains itself.